ktor-client/ktor-client-cio/jvmAndNix/src/io/ktor/client/engine/cio/Endpoint.kt

```diff
@@ -256,7 +257,18 @@ public class FailToConnectException : Exception("Connect timed out or retry atte
 
 internal expect fun Throwable.mapToKtor(request: HttpRequestData): Throwable
 
+@OptIn(InternalAPI::class)
 internal fun getRequestTimeout(
-    requestConfig: HttpTimeout.HttpTimeoutCapabilityConfiguration?,
+    request: HttpRequestData,
     engineConfig: CIOEngineConfig
-): Long = requestConfig?.requestTimeoutMillis ?: engineConfig.requestTimeout
+): Long {
+    /**
+     * The request timeout is handled by the plugin and disabled for the WebSockets.
+     */
+    val isWebSocket = request.url.protocol.isWebsocket()
+    if (request.getCapabilityOrNull(HttpTimeout) != null || isWebSocket || request.isUpgradeRequest()) {
```

Comment:
can you explain why we do `request.getCapabilityOrNull(HttpTimeout) != null` here?

Reply:
`CIO` engine had timeouts before we had `HttpTimeout` plugin. This place handles only case when the plugin is not installed.